src/stream/src/executor/merge.rs

```diff
@@ -200,24 +218,84 @@ impl Executor for MergeExecutor {
     }
 }
 
+#[derive(Default)]
+struct StagedWatermarks {
+    in_heap: bool,
+    staged: VecDeque<Watermark>,
+}
+
 pub struct SelectReceivers {
-    blocks: Vec<BoxedInput>,
-    upstreams: Vec<BoxedInput>,
+    blocks: Vec<(usize, BoxedInput)>,
+    upstreams: Vec<(usize, BoxedInput)>,
     barrier: Option<Barrier>,
     last_base: usize,
     actor_id: u32,
+    first_untransferred_watermarks: HashMap<usize, BinaryHeap<Reverse<(Watermark, usize)>>>,
+    other_untransferred_watermarks: HashMap<usize, HashMap<usize, StagedWatermarks>>,
```

Comment:
We won't have a lot of watermark columns usually. Using a `BTreeMap` here might be more efficient.

Reply:
Please also add some docs to these new fields and some comments to the code, as they seem not to be trivial.